src/reader.rs

```diff
@@ -3329,6 +3377,69 @@ impl<'a> Reader<'a> {
             }
         }
     }
+
+    fn _backward_token(&mut self) -> Option<(usize, EditableLineTag, usize)> {
+        let (elt, el) = self.active_edit_line();
+        let start_buff_pos = el.position();
+        let boundary = 0;
+        if start_buff_pos == boundary {
+            return None;
+        }
+
+        let buff_pos = el.position()
+            - el.text()[..el.position()]
+                .chars()
+                .take_while(|c| !c.is_alphanumeric())
+                .count()
+            - 1;
+
+        self.update_buff_pos(elt, Some(buff_pos));
+        let (elt, el) = self.active_edit_line();
+        let text = el.text();
+
+        let mut tok = 0..0;
+        let mut prev_tok = 0..0;
+        parse_util_token_extent(text, el.position(), &mut tok, Some(&mut prev_tok));
+
+        let new_position = if tok.start == start_buff_pos {
+            tok.start - 1 // this will never equal/exceed boundary
+        } else {
+            tok.start
+        };
+
+        Some((start_buff_pos, elt, new_position))
+    }
+
+    fn _forward_token(&mut self) -> Option<(usize, EditableLineTag, usize)> {
+        let (elt, el) = self.active_edit_line();
+        let start_buff_pos = el.position();
+        let boundary = el.len();
+        if start_buff_pos == boundary {
+            return None;
+        }
+
+        let buff_pos = el.position()
+            + el.text()[el.position()..]
+                .chars()
+                .take_while(|c| !c.is_alphanumeric())
```

Comment:
I don't think alphanumeric is right because it treats `_` differently for example.
Can we not parse without this offset and then use `prev_tok` instead of `tok`?

Reply:
hmm I copied most of this code from `transpose_words` but I'll try replacing `tok` with `prev_tok` instead of this and see if it still works

Reply:
If I remove `let buff_pos` and `self.update_buff_pos` from both functions...

```
        let buff_pos = el.position()
            + el.text()[el.position()..]
                .chars()
                .take_while(|c| !c.is_alphanumeric())
                .count();

        self.update_buff_pos(elt, Some(buff_pos));
```

Everything still works but when the cursor starts at a space then it only moves forward a single space

But you are right. The behavior in this case:

```
 _ '3 4'
```

is undesired. `ForwardToken` moves from `_` to `'`. 

I think we could replace `!c.is_alphanumeric()` with `c.is_whitespace()`. I will do that